File: test/ch21-benchmarks/bench-006-file-processing.go
// BENCH-006: File Line Processing - Go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func countErrorLines(filename string) (int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return 0, err
	}
	defer file.Close()

	count := 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if strings.Contains(strings.ToLower(scanner.Text()), "error") {
			count++
		}
	}

	if err := scanner.Err(); err != nil {
		return 0, err
	}

	return count, nil
}

func main() {
	result, err := countErrorLines("testdata/bench-006-logs-100mb.txt")
	if err != nil {
		panic(err)
	}
	fmt.Println(result)
	// Expected: 126076
}

File: test/ch21-benchmarks/bench-006-file-processing.c
// BENCH-006: File Line Processing - C
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int count_error_lines(const char *filename) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        perror("fopen");
        return -1;
    }

    int count = 0;
    char line[1024];
    
    while (fgets(line, sizeof(line), file)) {
        // Convert to lowercase and check for "error"
        char lower[1024];
        int i;
        for (i = 0; line[i] && i < 1023; i++) {
            lower[i] = tolower(line[i]);
        }
        lower[i] = '\0';
        
        if (strstr(lower, "error")) {
            count++;
        }
    }
    
    fclose(file);
    return count;
}

int main() {
    int result = count_error_lines("testdata/bench-006-logs-100mb.txt");
    printf("%d\n", result);
    // Expected: 126076
    return 0;
}
